hypothesis-python/src/hypothesis/searchstrategy/strategies.py

```diff
@@ -624,6 +624,41 @@ def do_draw(self, data):
                 # As long as we consume data, we'll eventually pass or raise.
                 # But if we don't this could be an infinite loop.
                 assume(data.index > start_index)
+
+        # Special-case for SampledFrom strategy.  This is special even among
+        # strategies with a known and small(ish) list of possible elements,
+        # because it implements drawing directly as choice of an index.
+        # That means we can use the "make your own luck" trick explained in
+        # the shrinking guide to find even very rare allowed values.
+        from hypothesis.searchstrategy.lazy import LazyStrategy
+        from hypothesis.searchstrategy.misc import SampledFromStrategy
+
+        while isinstance(self.filtered_strategy, LazyStrategy):
+            self.filtered_strategy = self.filtered_strategy.wrapped_strategy
```

Comment:
This is certainly a much more direct solution to the lazy-wrapper problem than mine (which involves a separate `LazyFilteredStrategy` class and some other supporting changes).

Note that there is an `unwrap_strategies` function in `lazy.py` that performs a similar nested unwrapping, though I can't say for sure whether it's applicable here.